File: Inventory_and_Order_Processing/Services/Inventory.cs
using DataStructures;
using Models;
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Services
{
    public class Inventory
    {
        private Dictionary<string, Product> _stock = new Dictionary<string, Product>();
        private CustomStack<Order> _undoStack = new CustomStack<Order>();
        private CustomQueue<Order> _shippingQueue = new CustomQueue<Order>();
        private string _ordersFilepath;

        public Inventory(string ordersFilepath)
        {
            _ordersFilepath = ordersFilepath;
        }
        public void LoadStock(string filepath)
        {
            if (!File.Exists(filepath))
            {
                Console.WriteLine($"Stock file not found: {filepath}");
                return;
            }

            foreach (string rawLine in File.ReadAllLines(filepath))
            {
                string line = rawLine.Trim();
                if (string.IsNullOrEmpty(line))
                {
                    continue;
                }

                string[] parts = line.Split(',');
                if (parts.Length != 3)
                {
                    Console.WriteLine($"Skipped malformed stock line: {line}");
                    continue;
                }

                string name = parts[0];
                if (!double.TryParse(parts[1], out double price) || !int.TryParse(parts[2], out int quantity))
                {
                    Console.WriteLine($"Skipped malformed stock line: {line}");
                    continue;
                }

                _stock[name] = new Product(name, price, quantity);
            }
        }

        public void LogOrder(Order order)
        {
            Directory.CreateDirectory(Path.GetDirectoryName(_ordersFilepath));
            using (StreamWriter writer = new StreamWriter(_ordersFilepath, true)) // append
            {
                writer.WriteLine($"Customer: {order.CustomerName} | Time: {order.Timestamp} | Total: {order.CalculateTotal()}");
                foreach (Product product in order.GetProducts())
                {
                    writer.WriteLine($"  {product.Name},{product.Price},{product.Quantity}");
                }
                writer.WriteLine(new string('=', 40));
            }
        }

        public void LogUndoNote(Order order)
        {
            Directory.CreateDirectory(Path.GetDirectoryName(_ordersFilepath));
            using (StreamWriter writer = new StreamWriter(_ordersFilepath, true)) // append
            {
                writer.WriteLine($"NOTE: Order for {order.CustomerName} (originally placed at {order.Timestamp}) was UNDONE.");
                writer.WriteLine(new string('=', 40));
            }
        }
        public Order PlaceOrder(string customerName, List<(string ProductName, int Quantity)> requestedItems)
        {
            foreach (var item in requestedItems)
            {
                if (!_stock.ContainsKey(item.ProductName))
                {
                    Console.WriteLine($"Order rejected: '{item.ProductName}' does not exist in stock.");
                    return null;
                }
                Product availableProduct = _stock[item.ProductName];
                if (item.Quantity > availableProduct.Quantity)
                {
                    Console.WriteLine($"Order rejected: not enough stock for '{item.ProductName}' " +
                                       $"(requested {item.Quantity}, available {availableProduct.Quantity}).");
                    return null;
                }
            }

            Order order = new Order(customerName);
            foreach (var item in requestedItems)
            {
                Product stockProduct = _stock[item.ProductName];
                stockProduct.ReduceQuantity(item.Quantity);

                Product orderedProduct = new Product(item.ProductName, stockProduct.Price, item.Quantity);
                order.AddProduct(orderedProduct);
            }

            _undoStack.Push(order);
            _shippingQueue.Enqueue(order);

            LogOrder(order);
            Console.WriteLine($"Order placed successfully for {customerName}.");
            return order;
        }

        public Order UndoLastOrder()
        {
            Order order = _undoStack.Pop();
            if (order == null)
            {
                return null;
            }

            foreach (Product product in order.GetProducts())
            {
                if (_stock.ContainsKey(product.Name))
                {
                    _stock[product.Name].IncreaseQuantity(product.Quantity);
                }
            }

            LogUndoNote(order);

            Console.WriteLine($"Order for {order.CustomerName} undone. Stock restored.");
            return order;
        }

        public Order ShipNextOrder()
        {
            Order order = _shippingQueue.Dequeue();
            if (order == null)
            {
                return null;
            }

            Console.WriteLine($"Shipping order for {order.CustomerName}:");
            order.DisplayInfo();
            return order;
        }

        public List<Product> GetLowStockProducts()
        {
            List<Product> lowStock = new List<Product>();
            foreach (Product product in _stock.Values)
            {
                if (product.Quantity < 5)
                {
                    lowStock.Add(product);
                }
            }
            return lowStock;
        }

        public List<Product> GetAllProducts()
        {
            return new List<Product>(_stock.Values);
        }
    }
}

File: Inventory_and_Order_Processing/inventory.py
import os
from models.product import Product
from models.order import Order
from data_structures.stack import Stack
from data_structures.queue import Queue


class Inventory:
    def __init__(self, orders_filepath):
        self._stock = {}
        self._undo_stack = Stack()
        self._shipping_queue = Queue()
        self._order_filepath = orders_filepath
        
    def load_stock(self, filepath):
        if not os.path.exists(filepath):
            print(f"Stock file not found: {filepath}")
            return
        
        with open(filepath, "r") as f:
            for raw_line in f:
                line = raw_line.strip()
                if not line:
                    continue
                
                parts = line.split(",")
                if len(parts) != 3:
                    print(f"Skipped malformed stock line: {line}")
                    continue
                
                name, price_str, quantity_str = parts
                try:
                    price = float(price_str)
                    quantity = int(quantity_str)
                except ValueError:
                    print(f"Skipped malformed stock line: {line}")
                    continue
                
                self._stock[name] = Product(name, price, quantity)
                
    def log_order(self, order):
        os.makedirs(os.path.dirname(self._order_filepath), exist_ok=True)
        with open(self._order_filepath, "a") as f:
            f.write(f"Customer: {order._customer_name} | Time: {order._timestamp} | Total: {order.calculate_total()}\n")
            for product in order._products:
                f.write(f" {product._name}, {product._price}, {product._quantity}\n")
            f.write(("=" * 40) + "\n")
            
    def log_undo_note(self, order):
        os.makedirs(os.path.dirname(self._order_filepath), exist_ok=True)
        with open(self._order_filepath, "a") as f:
            f.write(f"NOTE: Order for {order._customer_name} (originally placed at {order._timestamp}) was UNDONE.\n")
            f.write(("=" * 40) + "\n")

    def place_order(self, cusomter_name, requested_items):
        for product_name, requested_qty in requested_items:
            if product_name not in self._stock:
                print(f"Order rejected: '{product_name}' does bot exist in stock.")
                return None
            available_product = self._stock[product_name]
            if requested_qty > available_product._quantity:
                print(f"Order rejected: not enough stock for '{product_name}'"
                      f"(requested {requested_qty}, available {available_product._quantity}).")
                return None
            
        order = Order(cusomter_name)
        for product_name, requested_qty in requested_items:
            stock_product = self._stock[product_name]
            stock_product._quantity -= requested_qty
            
            ordered_product = Product(product_name, stock_product._price, requested_qty)
            order.add_product(ordered_product)
            
        self._undo_stack.push(order)
        self._shipping_queue.enqueue(order)
        
        print(f"Order placed successfully for {cusomter_name}.")
        return order
    
    def undo_last_order(self):
        order = self._undo_stack.pop()
        if order is None:
            return None
        
        for product in order._products:
            if product._name in self._stock:
                self._stock[product._name]._quantity += product._quantity
        
        self.log_undo_note(order)
        print(f"Order for {order._customer_name}:")
        # order.display_info()
        return order
    
    def ship_next_order(self):
        order = self._shipping_queue.dequeue()
        if order is None:
            return None
        
        print(f"Shipping order for {order._customer_name}")
        order.display_info()
        return order
    
    def get_low_stock_products(self):
        low_stock =[]
        for product in self._stock.values():
            if product._quantity < 5:
                low_stock.append(product)
        return low_stock
    
    def get_all_products(self):
        return list(self._stock.values())